src/shared/bootspec.c

```diff
@@ -883,30 +875,38 @@ static int insert_boot_entry_addon(
                 char *location,
                 char *cmdline) {
 
-        if (!GREEDY_REALLOC(addons->items, addons->count + 1))
+        if (!GREEDY_REALLOC(addons->items, addons->n_items + 1))
                 return log_oom();
 
-        addons->items[addons->count] = (BootEntryAddon) {
+        addons->items[addons->n_items] = (BootEntryAddon) {
                 .location = location,
                 .cmdline = cmdline,
         };
-        addons->count++;
+        addons->n_items++;
 
         return 0;
 }
 
+static void boot_entry_addons_done(BootEntryAddons *addons) {
+        FOREACH_ARRAY(addon, addons->items, addons->n_items) {
+                free(addon->cmdline);
+                free(addon->location);
+        }
```

Comment:
one more thing, please set `addons->n_items = 0` and `addons->items = mfree(addons->items);`

i.e. we prefer that "shallow" destructors which after all leave the primary data structure around leave it around in a "safe" way, i.e. that calling the shallow destructor again is a NOP, and the structure does not contain invalidated pointers.

Reply:
(i mean, this leaks the array otherwise too)